topology/static.go

```diff
@@ -0,0 +1,161 @@
+}
+
+// NewStaticTopologyType creates a new static topology type
+func NewStaticTopologyType(opts m3db.StaticTopologyTypeOptions) m3db.TopologyType {
+	return &staticTopologyType{opts}
+}
+
+func (t *staticTopologyType) Create() (m3db.Topology, error) {
+	if err := t.opts.Validate(); err != nil {
+		return nil, err
+	}
+	return newStaticTopology(t.opts), nil
+}
+
+type staticTopology struct {
+	topologyMap staticTopologyMap
+}
+
+func newStaticTopology(opts m3db.StaticTopologyTypeOptions) m3db.Topology {
+	return &staticTopology{newStaticTopologyMap(opts)}
+}
+
+func (t *staticTopology) Get() m3db.TopologyMap {
+	return &t.topologyMap
+}
+
+func (t *staticTopology) GetAndSubscribe(ch chan<- m3db.TopologyMap) m3db.TopologyMap {
+	// Topology is static, ignore the subscription channel
+	return &t.topologyMap
+}
+
+func (t *staticTopology) Close() error {
+	return nil
+}
+
+type staticTopologyMap struct {
+	shardScheme         m3db.ShardScheme
+	orderedHosts        []m3db.Host
+	hostsByShard        [][]m3db.Host
+	orderedHostsByShard [][]orderedHost
+	replicas            int
+	quorum              int
+}
+
+func newStaticTopologyMap(opts m3db.StaticTopologyTypeOptions) staticTopologyMap {
+	totalShards := len(opts.GetShardScheme().All().Shards())
+	hostShardSets := opts.GetHostShardSets()
+	topoMap := staticTopologyMap{
+		shardScheme:         opts.GetShardScheme(),
+		orderedHosts:        make([]m3db.Host, 0, len(hostShardSets)),
+		hostsByShard:        make([][]m3db.Host, totalShards),
+		orderedHostsByShard: make([][]orderedHost, totalShards),
+		replicas:            opts.GetReplicas(),
+		quorum:              quorum(opts.GetReplicas()),
+	}
+
+	for idx, hostShardSet := range hostShardSets {
+		host := hostShardSet.Host()
+		topoMap.orderedHosts = append(topoMap.orderedHosts, host)
+		for _, shard := range hostShardSet.ShardSet().Shards() {
+			topoMap.hostsByShard[shard] = append(topoMap.hostsByShard[shard], host)
+			topoMap.orderedHostsByShard[shard] = append(topoMap.orderedHostsByShard[shard], orderedHost{
+				idx:  idx,
+				host: host,
+			})
+		}
+	}
+
+	return topoMap
+}
+
+type orderedHost struct {
+	idx  int
+	host m3db.Host
+}
+
+func (t *staticTopologyMap) Hosts() []m3db.Host {
+	return t.orderedHosts
+}
+
+func (t *staticTopologyMap) HostsLen() int {
+	return len(t.orderedHosts)
+}
+
+func (t *staticTopologyMap) ShardScheme() m3db.ShardScheme {
+	return t.shardScheme
+}
+
+func (t *staticTopologyMap) Route(id string) (uint32, []m3db.Host, error) {
+	shard := t.shardScheme.Shard(id)
+	if int(shard) >= len(t.hostsByShard) {
+		return shard, nil, errUnownedShard
+	}
+	return shard, t.hostsByShard[shard], nil
+}
+
+func (t *staticTopologyMap) RouteForEach(id string, shardFn m3db.RouteShardFn, forEachFn m3db.RouteForEachFn) error {
+	shard := t.shardScheme.Shard(id)
+	shardFn(shard)
+	return t.RouteShardForEach(shard, forEachFn)
+}
+
+func (t *staticTopologyMap) RouteShard(shard uint32) ([]m3db.Host, error) {
```

Comment:
Yeah just trying to provide symmetric API to `Route` and `RouteForEach` and `RouteShardForEach` definitely does get used.